restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/CallManager.java

```diff
@@ -502,46 +527,59 @@ private void invite(final Object message) throws IOException, NumberParseExcepti
+                                    sendNotification(errMsg, 11001, "warning", true);
+                                    final SipServletResponse resp = request.createResponse(SC_NOT_FOUND, "Cannot complete P2P call");
+                                    resp.send();
+                                }
+                            } else {
+                                //Extensions didn't allowed this call
+                                if (logger.isDebugEnabled()) {
+                                    final String errMsg = "Client not Allowed to make this outbound call";
+                                    logger.debug(errMsg);
+                                }
+                                String errMsg = "Cannot Connect to Client: " + toClient.getFriendlyName()
+                                        + " : Make sure the Client exist or is registered with Restcomm";
+                                sendNotification(errMsg, 11001, "warning", true);
+                                final SipServletResponse resp = request.createResponse(SC_FORBIDDEN, "Call not allowed");
+                                resp.send();
+                            }
+                            ec.executePostOutboundAction(er, extensions);
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
                         }
-                        // if all goes well with proxying the invitation on to the next client
-                        // then we can end further processing of this INVITE
-                    } else {
-                        String errMsg = "Cannot Connect to Client: " + toClient.getFriendlyName()
-                                + " : Make sure the Client exist or is registered with Restcomm";
-                        sendNotification(errMsg, 11001, "warning", true);
-                        final SipServletResponse resp = request.createResponse(SC_NOT_FOUND, "Cannot complete P2P call");
-                        resp.send();
-                    }
-                } else {
-                    //Extensions didn't allowed this call
-                    if (logger.isDebugEnabled()) {
-                        final String errMsg = "Client not Allowed to make this outbound call";
-                        logger.debug(errMsg);
                     }
-                    String errMsg = "Cannot Connect to Client: " + toClient.getFriendlyName()
-                            + " : Make sure the Client exist or is registered with Restcomm";
-                    sendNotification(errMsg, 11001, "warning", true);
-                    final SipServletResponse resp = request.createResponse(SC_FORBIDDEN, "Call not allowed");
-                    resp.send();
-                }
-                ec.executePostOutboundAction(er, this.extensions);
+                }, system.dispatcher());
```

Comment:
@agafox, system.dispatcher is used for all Akka operations. If for any reason the Push Notification operation blocks or takes a long time to complete, this task will block the threads that should be used for Akka operations only.
This happened when we used the Akka thread pool to upload the Recordings to Amazon S3. For a reason the upload was taking a long time and we end up to a thread starvation and thus no thread was available for Akka actors to take a new call for example.
For this reason, we provided a new Execution Context: https://github.com/RestComm/Restcomm-Connect/blob/2bb19ef41310bd8e6c1f87d094e5150c05ace5d1/restcomm/restcomm.application/src/main/resources/application.conf#L34. Check here how we use it: https://github.com/RestComm/Restcomm-Connect/blob/2bb19ef41310bd8e6c1f87d094e5150c05ace5d1/restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/mybatis/MybatisRecordingsDao.java#L83